### PR TITLE
{math}[foss/2019a] OR-Tools v7.1 w/ Python 3.7.2

### PR DESCRIPTION
**Pull request by wpoely86**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
login2.cerberus.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6126 CPU @ 2.60GHz, Python 2.7.5
See https://gist.github.com/67e67dcfe43ddca89609ee7c68440966 for a full test report.

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+diff --git a/makefiles/Makefile.python.mk b/makefiles/Makefile.python.mk
+index 2b548e0..b7f72ae 100755
+--- a/makefiles/Makefile.python.mk
++++ b/makefiles/Makefile.python.mk
+@@ -1113,7 +1113,8 @@ endif # ifneq ($(PYTHON_EXECUTABLE),)
+ 
+ .PHONY: install_python # Install Python OR-Tools on the host system
+ install_python: pypi_archive
+-	cd "$(PYPI_ARCHIVE_TEMP_DIR)$Sortools" && "$(PYTHON_EXECUTABLE)" setup.py install --user
++	mkdir -p $(DESTDIR)/lib/python3.7/site-packages
```
Comment (boegel) [inline easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch]: @wpoely86 Can we avoid hardcoding the `3.7` here?

Maybe use `$(UNIX_PYTHON_VER)` here, and define `UNIX_PYTHON_VER` using `$EBVERSIONPYTHON`?

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
++	cd "$(PYPI_ARCHIVE_TEMP_DIR)$Sortools" && PYTHONPATH=$(DESTDIR)/lib/python3.7/site-packages:$(PYTHONPATH) "$(PYTHON_EXECUTABLE)" setup.py install --prefix $(DESTDIR)
```
Comment (boegel) [inline easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch]: does it work with `pip`?

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+# wpoely86@gmail.com
+diff -Nur or-tools-7.1.orig/Makefile.local.eb or-tools-7.1/Makefile.local.eb
+--- or-tools-7.1.orig/Makefile.local.eb	1970-01-01 01:00:00.000000000 +0100
++++ or-tools-7.1/Makefile.local.eb	2019-05-20 20:49:11.941360942 +0200
+@@ -0,0 +1,5 @@
++UNIX_SWIG_BINARY = $(EBROOTSWIG)/bin/swig
++JAVA_HOME = $(EBROOTJAVA)
++UNIX_PYTHON_VER = 3.7
++UNIX_GUROBI_DIR=$(EBROOTGUROBI)
++GUROBI_LIB_VERSION=$(shell echo $(EBVERSIONGUROBI) | sed 's:\([0-9]\+\)\.\([0-9]\+\)\..*:\1\2:')
```
Comment (boegel) [inline easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch]: Sort these alphabetically (easier for humans to digest)?

But is adding this `Makefile.local.eb` and making the `Makefile` include it actually needed though?

Can't you just define these via `buildopts`, like this:

```python
buildopts = "UNIX_SWIG_BINARY=${EBROOTSWIG}/bin/swig JAVA_HOME=${EBROOTJAVA} ..."
```

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+diff --git a/ortools/linear_solver/gurobi_interface.cc b/ortools/linear_solver/gurobi_interface.cc
+index dcc993e..de9adbd 100644
+--- a/ortools/linear_solver/gurobi_interface.cc
++++ b/ortools/linear_solver/gurobi_interface.cc
+@@ -175,7 +175,7 @@ class GurobiInterface : public MPSolverInterface {
+   GRBenv* env_;
+   bool mip_;
+   int current_solution_index_;
+-  MPCallback* callback_ = nullptr;
++//  MPCallback* callback_ = nullptr;
```
Comment (boegel) [inline easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch]: What's this about? Also mention this in the patch description?

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-foss-2019a-Python-3.7.2.eb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+source_urls = [GITHUB_SOURCE]
+sources = ['v%(version)s.tar.gz']
+patches = ['%(name)s-%(version)s-fix-makefiles.patch']
+checksums = [
+    '6118a931ffcf172c7e2371c3392f82380e9d2a29840daba7caff239ad70d4b44',  # v7.1.tar.gz
+    '35d05c0c57f0691f7570b3560443ed7d933fc97c13dc18c2227e2a251c127414',  # OR-Tools-7.1-fix-makefiles.patch
+]
+
+github_account = 'google'
```
Comment (boegel) [inline easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-foss-2019a-Python-3.7.2.eb]: This should probably go above `source_urls`?

Doesn't matter w.r.t. functionality of course, just a style thing.

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-foss-2019a-Python-3.7.2.eb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    ('Python', '3.7.2'),
+    ('zlib', '1.2.11'),
+    ('Java', '1.8', '', True),
+    # Gurobi is optional
+    ('Gurobi', '8.1.1', '', True),
+]
+
+skipsteps = ['configure']
+
+prebuildopts = 'make third_party &&'
```
Comment (boegel) [inline easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-foss-2019a-Python-3.7.2.eb]: What's the `third_party` make target for exactly? Maybe clarify with a comment?

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-foss-2019a-Python-3.7.2.eb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+buildopts = 'python cc'
+
+preinstallopts = 'prefix="" DESTDIR=%(installdir)s'
```
Comment (boegel) [inline easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-foss-2019a-Python-3.7.2.eb]: This looks weird, can you clarify the difference between `prefix` and `DESTDIR` (in a comment on the line above)?

Also, I find this easier to read (but that's maybe personal preference):

```python
preinstallopts = "prefix='' DESTDIR=%(installdir)s"

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch (hunk 2 above)
Comment (wpoely86) [reply easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch]: no

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch (hunk 3 above)
Comment (wpoely86) [reply easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch]: they will not overwrite the contents of `Makefile.local`

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 2 out of 2 (1 easyconfigs in this PR)
login2.cerberus.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6126 CPU @ 2.60GHz, Python 2.7.5
See https://gist.github.com/a281fd726955e8b1a6fae9e0c4180c77 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
login2.cerberus.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6126 CPU @ 2.60GHz, Python 2.7.5
See https://gist.github.com/dbd2ce51d4c83ac7f6d39280ee008f4d for a full test report.

**Comment by wpoely86** [comment]
@boegel remarks fixed

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch (hunk 3 above)
Comment (boegel) [reply easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch]: Hmm, that surprises me to be honest. Did you figure out exactly why that doesn't work?

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch (hunk 1 above)
Comment (boegel) [reply easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch]: `3.7` is still hardcoded?

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch (hunk 3 above)
Comment (wpoely86) [reply easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch]: It's changed.

File: easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch (hunk 1 above)
Comment (wpoely86) [reply easybuild/easyconfigs/o/OR-Tools/OR-Tools-7.1-fix-makefiles.patch]: oeps, forgot to change. fixed now

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 1 out of 2 (1 easyconfigs in this PR)
node3115.skitty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 2.7.5
See https://gist.github.com/cab992a065c80f2885f45998c4e289da for a full test report.

**Comment by wpoely86** [comment]
@boegel it fails because you don't have a license for Gurobi:

```
RuntimeError: 10009, Request denied: license not valid for Gurobi version 8
```

**Comment by boegel** [comment]
@wpoely86 Yes, I'm working on getting that fixed.

Maybe someone else who has access to Gurobi licenses can test this in the mean time?

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node3109.skitty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 2.7.5
See https://gist.github.com/122f9e9f993cbb2a53c70b94ddf1db3b for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node2081.delcatty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/1eeb07d8e6b3fe68ee605df710fba23d for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node2419.golett.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/8d02d0291bd0135407083b2ddef1aa6b for a full test report.

**Comment by boegel** [comment]
Going in, thanks @wpoely86!

**Comment by boegel** [comment]
@wpoely86 Merging this broke the tests because we now have both `Java/11` (for `R-3.6.0-foss-2019a.eb` and `Unicycler-0.4.7-foss-2019a-Python-3.7.2.eb`) and `Java/8` for `OR-Tools`...

Can we bump the `Java` dep for `OR-Tools` to `Java/11` to fix that?

**Pull request merged**